### ISSUE TITLE
When I use kubernetes-internal-ip as IP_AUTODETECTION_METHOD, router id generates error.

### ISSUE THREAD

**Issue by Roronoa-Z3r0** [projectcalico/calico#5851]
## Expected Behavior
Use kubeadm config for k8s initialization, the configuration is as follows，two node environment
kubernetes version 1.23.4
calico version 3.22.1
```
apiServer:
  certSANs: []
  extraArgs:
    enable-admission-plugins: NodeRestriction,DefaultStorageClass
apiVersion: kubeadm.k8s.io/v1beta2
etcd:
  local:
    dataDir: /var/lib/etcd
    extraArgs:
      election-timeout: '5000'
      heartbeat-interval: '500'
imageRepository: localhost
kind: ClusterConfiguration
kubernetesVersion: v1.23.4
networking:
  serviceSubnet: 10.96.0.0/12
---
apiVersion: kubeadm.k8s.io/v1beta2
bootstrapTokens:
- description: another bootstrap token
  token: 783bde.3f89s0fje9f38fhf
certificateKey: e6a2eb8581237ab72a4f494f30285ec12a9694d750b9785706a83bfcbbbd2204
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: 10.4.71.157
nodeRegistration:
  criSocket: /var/run/dockershim.sock
  kubeletExtraArgs:
    allowed-unsafe-sysctls: net.core.somaxconn
    node-ip: 10.4.71.157
  taints: []
---
apiVersion: kubelet.config.k8s.io/v1beta1
evictionHard:
  imagefs.available: 0%
  nodefs.available: 0%
imageGCHighThresholdPercent: 100
imageGCLowThresholdPercent: 99
kind: KubeletConfiguration

```
calico is configured as follows
```
cni_network_config: |-
    {
      "name": "k8s-pod-network",
      "cniVersion": "0.3.1",
      "plugins": [
        {
          "type": "calico",
          "log_level": "info",
          "log_file_path": "/var/log/calico/cni/cni.log",
          "datastore_type": "kubernetes",
          "nodename": "__KUBERNETES_NODE_NAME__",
          "mtu": __CNI_MTU__,
          "ipam": {

              "assign_ipv4": "true",


              "assign_ipv6": "false",

              "type": "calico-ipam"
          },
          "policy": {
              "type": "k8s"
          },
          "kubernetes": {
              "kubeconfig": "__KUBECONFIG_FILEPATH__"
          }
        },
        {
          "type": "portmap",
          "snat": true,
          "capabilities": {"portMappings": true}
        },
        {
          "type": "bandwidth",
          "capabilities": {"bandwidth": true}
        }
      ]
    }

```
```
- name: CALICO_NETWORKING_BACKEND
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: calico_backend
            # Cluster type to identify the deployment type
            - name: CLUSTER_TYPE
              value: "k8s,bgp"
            # Auto-detect the BGP IP address.
            - name: IP
              value: "autodetect"
            - name: IP_AUTODETECTION_METHOD
              value: "kubernetes-internal-ip"
            # Enable IPIP
            - name: CALICO_IPV4POOL_IPIP
              value: "CrossSubnet"
            # Enable or Disable VXLAN on the default IP pool.
            - name: CALICO_IPV4POOL_VXLAN
              value: "Never"
            # Set MTU for tunnel device used if ipip is enabled
            - name: FELIX_IPINIPMTU
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: veth_mtu
            # Set MTU for the VXLAN tunnel device.
            - name: FELIX_VXLANMTU
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: veth_mtu
            # Set MTU for the Wireguard tunnel device.
            - name: FELIX_WIREGUARDMTU
              valueFrom:
                configMapKeyRef:
                  name: calico-config
                  key: veth_mtu

```
## Current Behavior
One of the pods running a pod error
```
[root@node-71-157 calico]# kubectl get pod -n kube-system
NAME                                       READY   STATUS             RESTARTS       AGE
calico-kube-controllers-66c6c56ccd-j2dfs   1/1     Running            0              15m
calico-node-c8tph                          0/1     CrashLoopBackOff   7 (4m7s ago)   15m
calico-node-mgf5v                          1/1     Running            0              15m
coredns-27spt                              1/1     Running            0              15m
etcd-node-71-157                           1/1     Running            0              15m
kube-apiserver-node-71-157                 1/1     Running            0              15m
kube-controller-manager-node-71-157        1/1     Running            47             15m
kube-proxy-rvvfh                           1/1     Running            0              15m
kube-proxy-twc5j                           1/1     Running            0              15m
kube-scheduler-node-71-157                 1/1     Running            94             15m
tiller-deploy-mdx2c                        1/1     Running            0              15m

```
error pod log
```
2022-04-06 14:01:34.583 [INFO][9] startup/startup.go 425: Early log level set to info
2022-04-06 14:01:34.583 [INFO][9] startup/utils.go 127: Using NODENAME environment for node name node-71-158
2022-04-06 14:01:34.583 [INFO][9] startup/utils.go 139: Determined node name: node-71-158
2022-04-06 14:01:34.583 [INFO][9] startup/startup.go 94: Starting node node-71-158 with version v3.22.1
2022-04-06 14:01:34.586 [INFO][9] startup/startup.go 430: Checking datastore connection
2022-04-06 14:01:34.601 [INFO][9] startup/startup.go 454: Datastore connection verified
2022-04-06 14:01:34.601 [INFO][9] startup/startup.go 104: Datastore is ready
2022-04-06 14:01:34.629 [INFO][9] startup/startup.go 483: Initialize BGP data
2022-04-06 14:01:34.629 [INFO][9] startup/autodetection_methods.go 225: Including CIDR information from host interface. CIDR="10.4.71.158/22"
2022-04-06 14:01:34.630 [INFO][9] startup/startup.go 559: Node IPv4 changed, will check for conflicts
2022-04-06 14:01:34.636 [WARNING][9] startup/startup.go 983: Calico node 'node-71-157' is already using the IPv4 address 10.4.68.0.
2022-04-06 14:01:34.636 [INFO][9] startup/startup.go 389: Clearing out-of-date IPv4 address from this node IP="10.4.68.0/22"
2022-04-06 14:01:34.653 [WARNING][9] startup/utils.go 49: Terminating
```
10.4.68.0 is not the ip on my network card, but my router

```
[root@node-71-157 calico]# route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         gateway         0.0.0.0         UG    0      0        0 ens160
10.4.68.0       0.0.0.0         255.255.252.0   U     0      0        0 ens160
link-local      0.0.0.0         255.255.0.0     U     1002   0        0 ens160
172.16.232.0    0.0.0.0         255.255.255.192 U     0      0        0 *
172.16.232.1    0.0.0.0         255.255.255.255 UH    0      0        0 cali5feb9483fc8
172.16.232.2    0.0.0.0         255.255.255.255 UH    0      0        0 cali87ce7b03bc4
172.16.232.3    0.0.0.0         255.255.255.255 UH    0      0        0 calicdb1a168eb9
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
```
bird configuration for a running pod
```
[root@node-71-157 /]# cat /etc/calico/confd/config/bird.cfg 
function apply_communities ()
{
}

# Generated by confd
include "bird_aggr.cfg";
include "bird_ipam.cfg";

router id 10.4.68.0;

# Configure synchronization between routing tables and kernel.
protocol kernel {
  learn;             # Learn all alien routes from the kernel
  persist;           # Don't remove routes on bird shutdown
  scan time 2;       # Scan kernel routing table every 2 seconds
  import all;
  export filter calico_kernel_programming; # Default is export none
  graceful restart;  # Turn on graceful restart to reduce potential flaps in
                     # routes when reloading BIRD configuration.  With a full
                     # automatic mesh, there is no way to prevent BGP from
                     # flapping since multiple nodes update their BGP
                     # configuration at the same time, GR is not guaranteed to
                     # work correctly in this scenario.
  merge paths on;    # Allow export multipath routes (ECMP)
}

```
## Steps to Reproduce (for bugs)

1.Initialize the cluster with kubeadm init --config=xxx
2.Deploy calico with kubectl apply -f calico.yaml
3.Join another node to the cluster
4.

## Your Environment

* Calico version 3.22.1
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes 1.23.4
* Operating System and version: centos 7.7
* Link to your project (optional):

**Comment by caseydavenport** [comment]
Sounds like the nodes might be selecting the wrong address - what do your Kubernetes Node objects have for their InternalIPs? That's what Calico is looking at to determine its IP.

**Comment by mgleung** [comment]
I am going to close this due to inactivity. @Roronoa-Z3r0 if you are still seeing this issue, feel free to reopen.

**Comment by hammer-83** [comment]
I'm seeing the exact same issue with "kubernetes-internal-ip". The environment is similar with Calico v3.22.2 vxlan, Ubuntu 20.04, kubeadm setup v1.23.6.

Here are various possibly relevant logs:

POD:
```
2022-05-10 20:44:37.733 [INFO][10] startup/startup.go 425: Early log level set to info
2022-05-10 20:44:37.734 [INFO][10] startup/utils.go 127: Using NODENAME environment for node name master03
2022-05-10 20:44:37.734 [INFO][10] startup/utils.go 139: Determined node name: master03
2022-05-10 20:44:37.734 [INFO][10] startup/startup.go 94: Starting node master03 with version v3.22.2
2022-05-10 20:44:37.737 [INFO][10] startup/startup.go 430: Checking datastore connection
2022-05-10 20:44:37.763 [INFO][10] startup/startup.go 454: Datastore connection verified
2022-05-10 20:44:37.764 [INFO][10] startup/startup.go 104: Datastore is ready
2022-05-10 20:44:37.804 [INFO][10] startup/startup.go 483: Initialize BGP data
2022-05-10 20:44:37.805 [INFO][10] startup/autodetection_methods.go 225: Including CIDR information from host interface. CIDR="192.168.56.13/24"
2022-05-10 20:44:37.805 [INFO][10] startup/startup.go 559: Node IPv4 changed, will check for conflicts
2022-05-10 20:44:37.811 [WARNING][10] startup/startup.go 983: Calico node 'master02' is already using the IPv4 address 192.168.56.0.
2022-05-10 20:44:37.811 [INFO][10] startup/startup.go 389: Clearing out-of-date IPv4 address from this node IP="192.168.56.0/24"
2022-05-10 20:44:37.829 [WARNING][10] startup/utils.go 49: Terminating
Calico node failed to start
```

NODE (redacted):
```
apiVersion: v1
kind: Node
metadata:
  annotations:
    kubeadm.alpha.kubernetes.io/cri-socket: /run/containerd/containerd.sock
    node.alpha.kubernetes.io/ttl: "0"
    projectcalico.org/IPv4VXLANTunnelAddr: 10.244.235.0
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: master03
    kubernetes.io/os: linux
    node-role.kubernetes.io/control-plane: ""
    node-role.kubernetes.io/master: ""
    node.kubernetes.io/exclude-from-external-load-balancers: ""
  name: master03
spec:
  podCIDR: 10.244.4.0/24
  podCIDRs:
  - 10.244.4.0/24
  taints:
  - effect: NoSchedule
    key: node-role.kubernetes.io/master
status:
  addresses:
  - address: 192.168.56.13
    type: InternalIP
  - address: master03
    type: Hostname
  nodeInfo:
    architecture: amd64
    containerRuntimeVersion: containerd://1.5.11
    kernelVersion: 5.4.0-109-generic
    kubeProxyVersion: v1.23.6
    kubeletVersion: v1.23.6
    osImage: Ubuntu 20.04.4 LTS
```

Node interfaces (ip addr redacted):
```
2: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 08:00:27:e0:50:aa brd ff:ff:ff:ff:ff:ff
    inet 192.168.56.13/24 brd 192.168.56.255 scope global dynamic enp0s8
       valid_lft 4961sec preferred_lft 4961sec
    inet6 fe80::a00:27ff:fee0:50aa/64 scope link
       valid_lft forever preferred_lft forever
```

NetPlan config (redacted):
```
network:
  ethernets:
    enp0s8:
      match:
        name: eth0
      set-name: enp0s8
      mtu: 1500
      dhcp4: true
      dhcp4-overrides:
        route-metric: 100
```

From the pod logs, it looks like the interface CIDR is transformed from 192.168.56.13/24 to 192.168.56.0 along the way somewhere. Since all the nodes are on the same subnet and configured the same way, they all end up with the same auto-detected IP 192.168.56.0. The reason I attempted to switch to this detection method was because the default one was not choosing the right interface (I have another one connecting all nodes to storage).

**Comment by antoinep92** [comment]
Hit with the same issue, but (at least for me) it happens also with `canReach` detection, so it's not specific to the internalIP / kubernetes detection method, but rather the handling of IP masks.
I'm not even sure why the mask is of interest for IP detection.

**Comment by rayburgemeestre** [comment]
Same in my case:

```
# kubectl get nodes -o jsonpath='{.items[*].status.addresses[0].address}'; echo
10.141.0.1 10.141.0.2 10.141.0.3 10.141.0.4 10.141.0.5 10.141.0.6 10.141.255.254
```

node003's calico-node will be the lucky one that comes up, and then, any of the others will run into this.

```
# kubectl logs -f -n kube-system                         calico-node-dfvt8
2022-05-30 12:23:19.085 [INFO][11] startup/startup.go 425: Early log level set to info
2022-05-30 12:23:19.085 [INFO][11] startup/utils.go 127: Using NODENAME environment for node name node001
2022-05-30 12:23:19.086 [INFO][11] startup/utils.go 139: Determined node name: node001
2022-05-30 12:23:19.086 [INFO][11] startup/startup.go 94: Starting node node001 with version v3.23.1
2022-05-30 12:23:19.087 [INFO][11] startup/startup.go 430: Checking datastore connection
2022-05-30 12:23:19.103 [INFO][11] startup/startup.go 454: Datastore connection verified
2022-05-30 12:23:19.103 [INFO][11] startup/startup.go 104: Datastore is ready
2022-05-30 12:23:19.132 [INFO][11] startup/startup.go 483: Initialize BGP data
2022-05-30 12:23:19.133 [INFO][11] startup/autodetection_methods.go 225: Including CIDR information from host interface. CIDR="10.141.0.1/16"
2022-05-30 12:23:19.133 [INFO][11] startup/startup.go 559: Node IPv4 changed, will check for conflicts
2022-05-30 12:23:19.144 [WARNING][11] startup/startup.go 984: Calico node 'node003' is already using the IPv4 address 10.141.0.0.
2022-05-30 12:23:19.144 [INFO][11] startup/startup.go 389: Clearing out-of-date IPv4 address from this node IP="10.141.0.0/16"
2022-05-30 12:23:19.155 [WARNING][11] startup/utils.go 49: Terminating
```

**Comment by remram44** [comment]
Duplicate of #6141, fixed by #6228
